File: user-pages/src/lib/payout.ts
import { getSessionToken } from '@descope/react-sdk'

const API_BASE_URL = import.meta.env.VITE_BACKEND_URL

export interface FundedAccountPayout {
  account_id: number;
  challenge_id: string;
  account_size: string;
  current_balance: number;
  available_payout: number;
  profit_cap_amount: number;
  profit_split_percent: number;
  minimum_withdrawal_amount: number;
  withdrawal_count: number;
  last_withdrawal_at: string | null;
}

export interface WithdrawalHistory {
  id: number
  amount: number
  status: string
  requested_at: string
  completed_at: string | null
  reference: string
}

export interface PayoutEligibility {
  is_eligible: boolean
  has_verified_bank_account: boolean
  has_available_payout: boolean
  minimum_payout_amount: number
  bank_account_masked: string | null
  ineligibility_reasons: string[]
}

export interface PayoutSummaryResponse {
  total_available_payout: number
  total_earned_all_time: number
  funded_accounts: FundedAccountPayout[]
  withdrawal_history: WithdrawalHistory[]
  eligibility: PayoutEligibility
}

export interface PayoutEligibilityResponse {
  eligible: boolean
  has_bank_account: boolean
  has_funded_accounts: boolean
  available_payout: number
  reasons: string[]
}

class PayoutAPI {
  private async request(endpoint: string, options: RequestInit = {}): Promise<any> {
    const token = getSessionToken()
    if (!token) {
      throw new Error('No authentication token available')
    }

    const url = `${API_BASE_URL}${endpoint}`
    const config: RequestInit = {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        ...options.headers,
      },
      ...options,
    }

    const response = await fetch(url, config)

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}))
      throw new Error(errorData.detail || `HTTP ${response.status}: ${response.statusText}`)
    }

    return response.json()
  }

  async getPayoutSummary(): Promise<PayoutSummaryResponse> {
    return this.request('/payout/summary')
  }

  async checkEligibility(): Promise<PayoutEligibilityResponse> {
    return this.request('/payout/eligibility')
  }

  async requestPayout(accountId: number): Promise<{ request_id: string; amount: number; status: string; estimated_completion: string; message: string }> {
    return this.request('/payout/request', {
      method: 'POST',
      body: JSON.stringify({
        account_id: accountId,
      }),
    })
  }
}

export const payoutAPI = new PayoutAPI()

// Utility functions for formatting
export const formatCurrency = (amount: number): string => {
  return new Intl.NumberFormat('en-NG', {
    style: 'currency',
    currency: 'NGN',
    minimumFractionDigits: 2,
  }).format(amount).replace('NGN', '₦')
}

export const formatDate = (dateString: string): string => {
  const date = new Date(dateString)
  return date.toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  })
}

export const formatTimeAgo = (dateString: string): string => {
  const now = new Date()
  const date = new Date(dateString)
  const diffInHours = Math.floor((now.getTime() - date.getTime()) / (1000 * 60 * 60))

  if (diffInHours < 1) {
    return 'Just now'
  } else if (diffInHours < 24) {
    return `${diffInHours}h ago`
  } else {
    const diffInDays = Math.floor(diffInHours / 24)
    return `${diffInDays}d ago`
  }
}

File: backend/app/schemas/payout.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class FundedAccountPayout(BaseModel):
    """Payout information for a single funded account"""
    account_id: int
    challenge_id: str
    account_size: str
    current_balance: float
    available_payout: float
    profit_cap_amount: float
    profit_split_percent: int
    minimum_withdrawal_amount: float
    withdrawal_count: int
    last_withdrawal_at: Optional[datetime]

    class Config:
        from_attributes = True


class WithdrawalHistory(BaseModel):
    """Withdrawal transaction history"""
    id: int
    amount: float
    status: str
    requested_at: datetime
    completed_at: Optional[datetime]
    reference: str

    class Config:
        from_attributes = True


class PayoutEligibility(BaseModel):
    """User payout eligibility information"""
    is_eligible: bool
    has_verified_bank_account: bool
    has_available_payout: bool
    minimum_payout_amount: float
    bank_account_masked: Optional[str] = None
    ineligibility_reasons: List[str] = Field(default_factory=list, description="Reasons why user is not eligible for payout")


class PayoutSummaryResponse(BaseModel):
    """Complete payout summary for user dashboard"""
    total_available_payout: float = Field(..., description="Total payout available across all funded accounts")
    total_earned_all_time: float = Field(..., description="Total earnings including withdrawn amounts")
    funded_accounts: List[FundedAccountPayout] = Field(..., description="Detailed payout info for each funded account")
    withdrawal_history: List[WithdrawalHistory] = Field(..., description="Recent withdrawal transactions")
    eligibility: PayoutEligibility = Field(..., description="Payout eligibility status")


class PayoutRequest(BaseModel):
    """Request to initiate a payout"""
    amount: Optional[float] = Field(None, gt=0, description="Amount to withdraw (optional; defaults to max allowed)")
    account_id: Optional[int] = Field(None, description="Specific account to withdraw from (optional)")
    pin: str = Field(..., min_length=4, max_length=4, pattern="^\\d{4}$", description="Transaction PIN")


class PayoutRequestResponse(BaseModel):
    """Response after initiating a payout request"""
    request_id: str
    amount: float
    status: str
    estimated_completion: Optional[datetime]
    message: str